src/xmipp/libraries/reconstruction/resolution_fso.cpp

```diff
@@ -25,1334 +25,1078 @@
-			}
-			mask.clear();
-			pmask.clear();
+			double valmask = (double) DIRECT_MULTIDIM_ELEM(pmask, n);
+			DIRECT_MULTIDIM_ELEM(half1, n) = DIRECT_MULTIDIM_ELEM(half1, n) * valmask;
+			DIRECT_MULTIDIM_ELEM(half2, n) = DIRECT_MULTIDIM_ELEM(half2, n) * valmask;
 		}
+	}
+	mask.clear();
+	pmask.clear();
 
-		phalf1.setXmippOrigin();
-		phalf2.setXmippOrigin();
 
-    	std::cout << "Starting..." << std::endl;
-    }
+	half1.setXmippOrigin();
+	half2.setXmippOrigin();
 
+	// fftshift must by applied before computing the fft. This will be computed later
+	CenterFFT(half1, true);
+	CenterFFT(half2, true);
+}
 
-    void saveFSCToMetadata(MetaData &mdRes,
-    		const MultidimArray<double> &freq,
-			const MultidimArray<double> &FSC, FileName &fnmd)
-    {
-    	size_t id;
-    	FOR_ALL_ELEMENTS_IN_ARRAY1D(freq)
-		{
-			if (i>0)
-			{
-				id=mdRes.addObject();
-				mdRes.setValue(MDL_RESOLUTION_FREQ,dAi(freq, i),id);
-				mdRes.setValue(MDL_RESOLUTION_FRC,dAi(FSC, i),id);
-				mdRes.setValue(MDL_RESOLUTION_FREQREAL,1./dAi(freq, i),id);
-			}
-		}
-		mdRes.write(fnmd);
-    }
 
-    void saveAnisotropyToMetadata(MetaData &mdAnisotropy,
-    		const MultidimArray<double> &freq,
-			const MultidimArray<double> &anisotropy, FileName &fnmd)
-    {
-    	size_t objId;
-		FOR_ALL_ELEMENTS_IN_ARRAY1D(anisotropy)
+void ProgFSO::saveAnisotropyToMetadata(MetaData &mdAnisotropy,
+		const MultidimArray<double> &freq,
+		const MultidimArray<double> &anisotropy)
+{
+	size_t objId;
+	FOR_ALL_ELEMENTS_IN_ARRAY1D(anisotropy)
+	{
+		if (i>0)
 		{
-			if (i>0)
-			{
-			objId = mdAnisotropy.addObject();
-			mdAnisotropy.setValue(MDL_RESOLUTION_FREQ, dAi(freq, i),objId);
-			mdAnisotropy.setValue(MDL_RESOLUTION_FRC, dAi(anisotropy, i),objId);
-			mdAnisotropy.setValue(MDL_RESOLUTION_FREQREAL, 1.0/dAi(freq, i),objId);
-			}
+		objId = mdAnisotropy.addObject();
+		mdAnisotropy.setValue(MDL_RESOLUTION_FREQ, dAi(freq, i),objId);
+		mdAnisotropy.setValue(MDL_RESOLUTION_FSO, dAi(anisotropy, i),objId);
+		mdAnisotropy.setValue(MDL_RESOLUTION_FREQREAL, 1.0/dAi(freq, i),objId);
 		}
-		mdAnisotropy.write(fnmd);
-    }
+	}
+	mdAnisotropy.write(fnOut+"/fso.xmd");
+}
 
 
-    // A directional filtered map is obtained by applying the 3dfsc as filter in Fourier domain
-    void directionalFilter(MultidimArray<std::complex<double>> &FThalf1,
-    		MultidimArray<double> &threeDfsc, MultidimArray<double> &filteredMap, int m1sizeX, int m1sizeY, int m1sizeZ)
+void ProgFSO::directionalFilter(MultidimArray<std::complex<double>> &FThalf1, 
+			MultidimArray<double> &threeDfsc, 
+			MultidimArray<double> &filteredMap, int m1sizeX, int m1sizeY, int m1sizeZ)
     {
-
-    	Image<double> imgHalf1;
+    	Image<double> imgHalf1, imgHalf2;
     	imgHalf1.read(fnhalf1);
-    	MultidimArray<double> half1;
+		imgHalf2.read(fnhalf2);
+    	MultidimArray<double> half1, half2;
     	half1 = imgHalf1();
+		half2 = imgHalf2();
 
-	//FIXME: Use atf.h
         FourierTransformer transformer1(FFTW_BACKWARD);
```

Comment:
use the `FFTwT`, and as a bonus you can use multiple threads and floats

Reply:
Iknow, let me keep it in this way, since now I will use atf library.